### PR TITLE
Add new Derived stream class

### PR DESCRIPTION
**Pull request by jonmmease**
closes https://github.com/holoviz/holoviews/issues/4530

This PR introduces a new `Derived` stream class that generalizes the concept of having a `Stream` that produces values based on one or more other input streams.  

The `SelectionExpr` stream had done this in a somewhat ad-hoc way, but with this PR `SelectionExpr` is now a subclass of the new `Derived` stream.

The existing `SelectionExpr` tests pass with minimal updates. I also added a new set of tests for `Derived` streams directly in `holoviews/tests/teststreams.py`.  This is a good place to look at for a simple example of what a `Derived` stream subclass looks like.

Let me know if you have any questions or concerns about the approach or API. Thanks!

### CODE REVIEW

File: holoviews/tests/teststreams.py [hunk 1]
```diff
@@ -889,14 +971,14 @@ def test_selection_expr_stream_scatter_points(self):
             expr_stream = SelectionExpr(element)
 
             # Check stream properties
-            self.assertEqual(len(expr_stream._source_streams), 2)
-            self.assertIsInstance(expr_stream._source_streams[0], SelectionXY)
-            self.assertIsInstance(expr_stream._source_streams[1], Lasso)
+            self.assertEqual(len(expr_stream.input_streams), 2)
```
Comment (jbednar) [inline holoviews/tests/teststreams.py:974]: Is there a particular reason to switch from `source` to `input` here?  Does that help make the terminology more consistent?

Comment (jonmmease) [reply holoviews/tests/teststreams.py:974]: Here was my thinking. For the `SelectionExpr` stream, it happened to be the case that all of the input streams were always attached to a single `source` element.  The `Derived` stream doesn't assume this, all it knows is that it should calculate a value based on a list of streams that come from somewhere else.  So I guess it seemed to me that using "source" here was no longer accurate.

Comment (jbednar) [reply holoviews/tests/teststreams.py:974]: Thanks; makes sense to me!

**Comment by jonmmease** [comment]
The AppVeyor tests are failing due to a bunch a bunch of failures in `spatialpandas` that I think are pyarrow 1.0 incompatibilities.  Though it looks like spatialpandas pins the pyarrow dependency to `<1` (https://github.com/holoviz/spatialpandas/blob/master/setup.py#L27), so I'm not sure how this is happening.

Travis tests aren't even being run I don't think:
```
$ pip install pyctdev && doit miniconda_install && pip uninstall -y doit pyctdev

pyenv: version `3.7' is not installed (set by PYENV_VERSION environment variable)

The command "pip install pyctdev && doit miniconda_install && pip uninstall -y doit pyctdev" failed and exited with 1 during .
```
Does this mean anything to anyone else?

**Comment by jbednar** [comment]
spatialpandas master pins to <1.0 to avoid those issues, but it hasn't been released yet, so you're presumably still getting an older spatialpandas here. Pinning pyarrow here seems reasonable to me; @jlstevens was looking at releasing spatialpandas.

**Comment by jbednar** [comment]
Oh, and this particular immediate issue is presumably that Travis is providing 3.6 or 3.8, while the rest of the setup is assuming 3.7.  That happens every time Travis updates their default images, I think, and we have to adapt. I forget the details, but again @jlstevens will presumably remember.

**Comment by philippjfr** [comment]
The pyenv stuff can be fixed by switching back to `language: python` in the travis.yml.

**Comment by philippjfr** [comment]
No objection to pinning pyarrow in the test dependencies in setup.py for now.

**Comment by jonmmease** [comment]
I went ahead and got spatialpandas working with pyarrow 1.0 in https://github.com/holoviz/spatialpandas/pull/46.  So hopefully this will take care of the issue here once that package is released.

**Comment by philippjfr** [comment]
Looking good, I'll review and hopefully merge this evening.

File: holoviews/tests/teststreams.py (hunk 1 above)
Comment (philippjfr) [reply holoviews/tests/teststreams.py:974]: Makes sense to me too.

**Comment by jonmmease** [comment]
Actually, I'm working on the `uncollate` transform and I may need to revisit this a little. The `trigger_index` idea doesn't really fit with the stateless goal, so I may need to think on how to remove this.

As you have time, I'd appreciate feedback on https://github.com/holoviz/holoviews/pull/4534 and I'll update this PR once I've gotten further with `uncollate`.  Thanks!

**Comment by jonmmease** [comment]
Ok, I removed `trigger_index` and added an `exclusive` option to `Derived` that, when set, will clear all streams except the most recently updated stream.

**Comment by jlstevens** [comment]
I know this is WIP and I am only raising a naming issue, but I don't think `uncollate` is a real word. :-)

The antonym I probably liked the most from [this list](https://www.thesaurus.com/browse/collate) is probably `separate` but I would like to hear any other suggestions!

**Comment by jbednar** [comment]
It may not be the best word, but it's a word people use: https://www.support.xerox.com/en-us/article/en/x_wc7245_en-O12395 

I think it's useful to have the visual semantic link to `collate`, so that they can be understood as duals. As long as that's true (i.e. that `collate(uncollate(x)) = x`, at least roughly), a different name would need to be quite compelling before it would unseat `uncollate`.  Conversely, if it's not really the dual of `collate`, then `uncollate` is confusing and shouldn't be used...

**Comment by jonmmease** [comment]
> I think it's useful to have the visual semantic link to collate, so that they can be understood as duals. As long as that's true (i.e. that collate(uncollate(x)) = x, at least roughly), a different name would need to be quite compelling before it would unseat uncollate

Yeah, I believe these should be very nearly duals. I'll document more when that PR is ready. 

@jlstevens, just to clarify, `uncollate` doesn't show up in this PR. I'll have another PR soon that introduces it. Thanks!

**Comment by philippjfr** [comment]
@jonmmease Could you rebase both PRs? CI is working again on master.

**Comment by jonmmease** [comment]
Updated with master, and things have gotten further!

In AppVeyor, seeing this error
```
  File "C:\Miniconda36-x64\envs\test\lib\distutils\version.py", line 337, in _cmp
    if self.version < other.version:
TypeError: '<' not supported between instances of 'str' and 'int'
```

with this in the stack trace:
```
C:\Miniconda36-x64\envs\test\lib\site-packages\panel\pane\holoviews.py in _render(self, doc, comm, root)
    278 
    279         kwargs = {'margin': self.margin}
--> 280         if backend == 'bokeh' or LooseVersion(str(hv.__version__)) >= str('1.13.0'):
    281             kwargs['doc'] = doc
    282             kwargs['root'] = root
C:\Miniconda36-x64\envs\test\lib\distutils\version.py in __ge__(self, other)
```

Based on https://bugs.python.org/issue14894, looks like this is an issue with comparing version numbers with different number parts that mix numbers and letters. e.g. this also raises the error `LooseVersion("22.7-r1") < LooseVersion("22.7.3-r1")`.

**Comment by philippjfr** [comment]
Let's fix appveyor separately, as long as Travis passes I'm happy to merge. Only a very minor issue there:

```
======================================================================
ERROR: holoviews.tests.teststreams.test_all_stream_parameters_constant
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/holoviz/holoviews/holoviews/tests/teststreams.py", line 28, in test_all_stream_parameters_constant
    % (name, stream_cls.__name__))
TypeError: Parameter v of stream Sum not declared constant
```

**Comment by jonmmease** [comment]
ping @philippjfr. Travis is green now. Thanks for taking the time to look this over!

**Comment by philippjfr** [comment]
Thank you for working on this!

**Pull request merged**